CIP-plutus-modules/README.md

```diff
@@ -0,0 +1,1656 @@
+with any implementations of `B` and `C`--the calling script must pass
+implementations of `B` and `C` to the lambda expression, and can
+choose them freely. We call this 'dynamic linking', because the
+implementation of dependencies may vary from use to use. On the other
+hand, if we want to *fix* the versions of `B` and `C` then we can
+create a `Script` that applies the same `CompleteScript` to two
+`ScriptArg`s, containing the hashes of the intended versions of `B`
+and `C`, which will then be supplied by
+`resolveScriptDependencies`. We call this 'static linking', because
+the version choice for the dependency is fixed by the script. It is up
+to script developers (or compiler writers) to decide between static
+and dynamic linking in this sense.
+
+On the other hand, when a script is used directly as a validator then
+there is no opportunity to supply additional arguments; all modules
+used must be supplied as `ScriptArg`s, which means they are
+fixed. This makes sense: it would be perverse if a transaction trying
+to spend a UTxO protected by a spending validator were allowed to
+replace some of the validation code--that would open a real can of
+worms, permitting many attacks whenever a script was split over
+several modules. With the design in the CIP, it is the script in the
+UTxO being spent that determines the module versions to be used, not
+the spending transaction. That transaction does need to *supply* all
+the modules actually used--including all of their dependencies--but it
+cannot choose to supply alternative implementations of them.
+
+### In-service upgrade
+
+Long-lived contracts may need upgrades and bug fixes during their
+lifetimes. This need is met on the Ethereum blockchain using the
+'proxy pattern'--a 'proxy' contract which delegates calls to the
+current implementation contract, whose identity is stored in the proxy
+contract's mutable state. Proxy contracts can provide a 'code upgrade'
+method which modifies the mutable state to store a new implementation
+contract.
+
+
+The Cardano chain does not offer mutable state. Instead, a changing
+state is represented by a succession of UTxOs, each holding the
+current state, usually with the currently-valid UTxO identified by
+holding a particular NFT. In the absence of mutable state a dependency
+cannot be updated just by changing a pointer, but scripts can still be
+upgraded by creating new values on the chain. The exact mechanism
+depends on the kind of script--and, often, on the original script
+developer preparing the ground for a later code change.
+
+Note that, on Ethereum, a proxy contract can be updated without
+changing its contract address---thanks to mutable state. On Cardano, a
+script address *is* the hash of its code; of course, changing the code
+will change the script address. It is very hard to see how that could
+possibly be changed without a fundamental redesign of Cardano. So the
+methods discussed below are different in nature from the Ethereum one:

```

Comment:
Thanks--so I think you're saying two things:

1. Upgrade-in-place can be done more simply than suggested here, by using proxy scripts that delegate verification to a script hash kept in a state UTxO,

2. Upgrade is a minefield, and should be avoided altogether.

Where (1) is concerned, simple is good! I wonder though if it doesn't require you to buy in to the "one verifier to rule them all" approach, where one staking validator checks all the spending and minting in the transaction. I realise that's a popular approach, but not the only possible approach.

Where (2) is concerned, I have a lot of sympathy with that view, but at the same time I don't expect all libraries to be bug-free when they are first released, so it's a natural question to ask "what should I do if a library I am depending on receives a bug fix update?" It seems a little unrealistic just to ignore the problem altogether. Of course, "get it right first time" is good advice, but hard to follow consistently.

Maybe you're saying essentially: upgrading to new versions is a problem that already exists, and existing 'solutions' are equally applicable once modules are introduced--so this CIP need not address the problem specifically. Even if, by supporting larger code drawn from multiple sources, it's likely to make the problem worse.

Reply:
> Maybe you're saying essentially: upgrading to new versions is a problem that already exists, and existing 'solutions' are equally applicable once modules are introduced--so this CIP need not address the problem specifically. Even if, by supporting larger code drawn from multiple sources, it's likely to make the problem worse.

Yes. I think smart contract upgrades is an incredibly complex problem and must be handled with extreme care. I think that this CIP should not even cover it, because the suggestions here can lead people to underestimate the severity of the problem and the care with which they must handle it. 

> what should I do if a library I am depending on receives a bug fix update?

Say that you upgrade to the new "bug fix version" what if it introduces a hidden backdoor or introduces other vulnerabilities? Importantly, the users who signed transactions and agreed to put their funds into your dApp, agreed for their funds to be secured by scripts to which they sent their funds, they did not consent for their funds to be secured by the new "bug fix script" which could introduce a backdoor to steal all their funds. Any introduction of non-manual upgradability leads to the possibility of a backdoor to drain liquidity (i.e. "upgrade" the library to a malicious contract that always succeeds if the transaction is signed by the malicious actor's pub key hash). That's why you should prefer either:

1. manual migration for upgrades - users must migrate their liquidity themselves to the new scripts.
2. Onchain Token Governance DAO based upgrades - use an onchain governance protocol, like Agora, except these protocols are very experimental on Cardano, so much so that even the creators of Agora do not use it for governance of their protocol.

Reply:
OK, thanks... I'm persuaded.